designs/isolate-cluster-upgrades.md

```diff
@@ -0,0 +1,221 @@
+
+### Workflow: how to use `observedGeneration` to only reconcile on changes
+This diagram represents the logic to follow in the EKS-A `Cluster` controller:
+
+```mermaid
+stateDiagram-v2
+    [*] --> Queue: Cluster reconcile request
+    
+    deleted: Cluster deleted?
+    delete: Reconcile delete
+    reconcile: Reconcile cluster
+    observedGeneration: generation == observedGeneration
+    updateGeneration: observedGeneration = generation
+    done: Reconciliation done?
+
+    Queue --> deleted
+    state deleted <<choice>> 
+        deleted --> observedGeneration: no
+        deleted --> delete: yes
+    state observedGeneration <<choice>>
+        observedGeneration --> reconcile: no
+        observedGeneration --> [*]: yes
+    delete --> done
+    reconcile --> done
+    state done <<choice>>
+        done --> updateGeneration: yes
+        done --> Queue: no
+    updateGeneration  --> [*]
+```
+
+We only perform the actual reconciliation logic (this is trying to change the state of the world) when `observedGeneration` is different than `generation`. This will only happen after the resource has changed.
+
+It's important to mention that we only update the value of `observedGeneration` when the reconciliation process is "done". Done here means that the state of the world is equal to the desired state: the cluster has been completely reconciled. This is to take into account the case where a full reconciliation takes multiple loops (for example, when the cluster is first created). Note that "reconcile not done" also includes errors.
+
+This might be slightly different of how other resources use the `observedGeneration` field, where the field is always set on every reconciliation loop. For multi state resources (like our `Cluster`) this is usually accompanied of a "state" indicator (note that `phases` is now discouraged in favor of `conditions`). However, we don't have that in our `Cluster`'s status. Although tempting to try to add this to the status, we encourage to tackle this problem separately and revisit the use of `observedGeneration` if necessary. Refer to the [alternatives](#alternative-solutions) section for other possible solutions.
+
+### Cluster vs cluster config: how to calculate an aggregated generation
+An astute reader might have realized that until now we have worked with a simplified model: where cluster reconciliation only needs to occur when the resource been reconciled `Cluster` object changes. Although this is true for most controllers in Kubernetes, our Cluster controller doesn't behave this way. Due to the particularities of our API and how we define a single entity (the cluster config) with multiple CRDs (all linked through the `Cluster` object), the Cluster controller reacts to changes not only in the `Cluster` but also in all its linked objects (we'll call them child objects for the rest of the document).
```

Comment:
I am confused by "when the resource been reconciled `Cluster` object changes". Are you saying it only needs to occur when `Cluster` object changes or when it has already been reconciled and then it changes?

Reply:
ah this is bad writing
I should say:
> An astute reader might have realized that until now we have worked with a simplified model: one where cluster reconciliation only needs to occur when the resource been reconciled (the `Cluster` in this case) changes.